Generalize elastic constant conversion

hugary1995 opened the pull request:
- Separate out conversion functions for elastic constants so that they could be reused elsewhere independent of the models
- Fix cross-referencing when declare nonlinear parameter -- same change applied to BufferStore
- Separate out generic converters for better code reuse: see `ElasticityConverter.h`
- Do not calculate derivatives if the input elastic constants are not nonlinear parameters
- Move all elasticity-related models into the `elasticity` subfolder

hugary1995 committed:
Improving elasticity tensors

hugary1995 committed:
Generalize elasticity tensor conversion

hugary1995 committed:
Remove some if: !cancelled() in workflow files

hugary1995 committed:
Use the same interface for LinearIsotropicElasticity; modify input files; also fix LinearCombination

hugary1995 committed:
Fix remaining tests; Use shear modulus in LinearIsotropicElasticJ2TrialStressUpdate

reverendbedford commented on include/neml2/models/solid_mechanics/elasticity/ElasticityTensor.h:
```diff
@@ -0,0 +1,188 @@
+{
+  OptionSet options = Derived::expected_options();
+
+  MultiEnumSelection type_selection({"LAME_FIRST_CONSTANT",
```
Honestly `"LAME_LAMBDA"` may be better.  This always makes me have to think about which is the first parameter...

reverendbedford commented on include/neml2/models/solid_mechanics/elasticity/ElasticityConverter.h:
```diff
@@ -0,0 +1,120 @@
+  void assert_ascending(const ConverterKey & ps) const
+  {
+    for (std::size_t i = 1; i < N; ++i)
+      neml_assert(
```
```suggestion
      neml_assert_debug(
```

I assume, as this is an internal error.

hugary1995 replied:
This is internal. But I think it doesn't hurt to make this check a non `_dbg` one, because it's only called in the constructor and doesn't affect run speed.

reverendbedford replied:
Okay.

hugary1995 committed:
LAME_LAMBDA

hugary1995 commented:
Don't merge yet, one last change tomorrow morning.

hugary1995 committed:
- Rename ElasticityTensor->ElasticityInterface
- Use ElasticityInterface in LinearIsotropicElasticJ2TrialStressUpdate
- Some algebraic optimization in conversion formulae
- Allow nonlinear parameter in LinearIsotropicElasticJ2TrialStressUpdate and LinearCombination
- Add default parameter names for cubic constants
